File: src/com/kingmang/lazurite/parser/AST/Expressions/AssignmentExpression.java
package com.kingmang.lazurite.parser.AST.Expressions;

import com.kingmang.lazurite.parser.AST.Statements.Statement;
import com.kingmang.lazurite.parser.AST.Accessible;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.Value;
import lombok.AllArgsConstructor;


public final class AssignmentExpression extends InterruptableNode implements Expression, Statement {

    public final Accessible target;
    public final BinaryExpression.Operator operation;
    public final Expression expression;

    public AssignmentExpression(BinaryExpression.Operator operation, Accessible target, Expression expr) {
        this.operation = operation;
        this.target = target;
        this.expression = expr;
    }
    @Override
    public void execute() {
        eval();
    }

    @Override
    public Value eval() {
        super.interruptionCheck();
        if (operation == null) {
            // Simple assignment
            return target.set(expression.eval());
        }
        final Expression expr1 = new ValueExpression(target.get());
        final Expression expr2 = new ValueExpression(expression.eval());
        return target.set(new BinaryExpression(operation, expr1, expr2).eval());
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        final String op = (operation == null) ? "" : operation.toString();
        return String.format("%s %s= %s", target, op, expression);
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/expressions/AssignmentExpression.kt
package com.kingmang.lazurite.parser.ast.expressions

import com.kingmang.lazurite.parser.ast.Accessible
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.statements.Statement
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.values.LzrValue

data class AssignmentExpression(
    val operation: BinaryExpression.Operator?,
    val target: Accessible,
    val expression: Expression
) : InterruptableNode(), Expression, Statement {
    override fun execute() {
        eval()
    }

    override fun eval(): LzrValue {
        super.interruptionCheck()
        if (operation == null)
            return target.set(expression.eval())
        val expr1: Expression = ValueExpression(target.get())
        val expr2: Expression = ValueExpression(expression.eval())
        return target.set(BinaryExpression(operation, expr1, expr2).eval())
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R =
        visitor.visit(this, t)

    override fun toString(): String =
        String.format("%s %s= %s", target,  if ((operation == null)) "" else operation.toString(), expression)
}
